tools/istio-iptables/pkg/builder/iptables_builder_impl.go

```diff
@@ -170,6 +170,45 @@ func (rb *IptablesRuleBuilder) buildRules(rules []*Rule) [][]string {
 	return output
 }
 
+func (rb *IptablesRuleBuilder) buildCleanupRules(rules []*Rule) [][]string {
```

Comment:
Nice! TY!

If all our rules are in custom chains, does this

- simply drop the entire chain on removal, as well as all jumps to the chain?
or
- run _exactly_ in reverse and remove every rule from the chain as well as the chain itself and all jumps?

the first is all we need really if we stick to the right conventions internally, but if this is an automated inversion of rules creation that doesn't rely on convention to reliably reverse any input, I wouldn't say no to it.

Reply:
Also, for init containers, upgrades are not a concern (AFAICT), so reversing your created rules exactly will always work for cleanup.

For things like ambient, we might need to contend with cleanup after an upgrade, where rules might have changed subtly between releases - so reversing the 1.21 hardcoded rules might not 100% match the reversed 1.20 rules, and such.

That's a scenario where I wonder if dropping chains and jumps with a given prefix might be more robust than strict mirror-reversal.

Reply:
(Sorry, forgot to respond here)



>     ISTIO_* Chain Jump Rule: If a rule is created to jump to an ISTIO_* chain, reverse it. This would cover also an hypothetical jump from an ISTIO_CHAIN to another ISTIO_CHAIN.
>     
>     Non-ISTIO_* Chain Rule: If a rule is created in a non-ISTIO_* chain, reverse it (should I care about this outside of the jump-rules to istio chains?) yes I should since we also apply those under certain circumstances
>     
>     ISTIO_* Chain Rule: If a rule is created inside an ISTIO_* chain, ignore it (it will be deleted later with the chain).
>     Delete ISTIO_* Chains: At the end, all ISTIO_* chains are deleted regardless.

This sounds good to me, for the following reasons.

1. `istio-iptables` is used by both `istio-init` sidecars and `ambient`, so new stuff we add must be something that can work for both.

2. I want to discourage people/ourselves from adding Istio rules to non-prefixed chains - in sidecar and ambient. 
We should never add anything to main tables but jumps. If we are, we should tag those with TODO and/or WARN on remove for those rules.

To facilitate #2, can you add a WARN log to the cleanup if it finds a non-jump rule in the main table, as a nag to us to eventually remove those?

Reply:
I will open an issue to fix them - I would rather have the nag than not :D

EDIT: https://github.com/istio/istio/issues/50532